sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingJoinSuite.scala

```diff
@@ -1041,3 +1077,104 @@ class StreamingOuterJoinSuite extends StreamTest with StateStoreMetricsTest with
     )
   }
 }
+
+class StreamingLeftSemiJoinSuite extends StreamingJoinSuite {
+
+  import testImplicits._
+
+  test("windowed left semi join") {
+    val (leftInput, rightInput, joined) = setupWindowedJoin("left_semi")
+
+    testStream(joined)(
+      MultiAddData(leftInput, 1, 2, 3, 4, 5)(rightInput, 3, 4, 5, 6, 7),
+      CheckNewAnswer(Row(3, 10, 6), Row(4, 10, 8), Row(5, 10, 10)),
+      MultiAddData(leftInput, 21)(rightInput, 22),
+      // Watermark = 11, should remove rows having window=[0,10]
+      CheckNewAnswer(),
+      assertNumStateRows(total = 2, updated = 12),
+      AddData(leftInput, 22),
+      CheckNewAnswer(Row(22, 30, 44)),
+      assertNumStateRows(total = 2, updated = 0),
+      StopStream,
+      StartStream(),
+
+      AddData(leftInput, 1),
+      // Row not add as 1 < state key watermark = 12
+      CheckNewAnswer(),
+      AddData(rightInput, 11),
+      // Row not add as 11 < state key watermark = 12
+      CheckNewAnswer()
+    )
+  }
+
+  test("left semi early state exclusion on left") {
+    val (leftInput, rightInput, joined) = setupWindowedJoinWithLeftCondition("left_semi")
+
+    testStream(joined)(
+      MultiAddData(leftInput, 1, 2, 3)(rightInput, 1, 2, 3, 4, 5),
+      // The left rows with leftValue <= 4 should not generate their semi join row and
+      // not get added to the state.
+      CheckNewAnswer(Row(3, 10, 6)),
+      assertNumStateRows(total = 4, updated = 4),
```

Comment:
Can you describe the details on the state rows, and add the same num state rows verification on below after each CheckNewAnswer?